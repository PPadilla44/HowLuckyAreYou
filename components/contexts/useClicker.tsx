import React, { createContext, Dispatch, ReactNode, useContext, useReducer } from 'react';
import { Action, initialClickerState, ClickerState } from '../../store/clicker';
import clickerReducer from "../../store"

const ClickerContext = createContext<{ state: ClickerState, dispatch?: Dispatch<Action>  }>({ state: initialClickerState });

export const useClicker = () => {
    return useContext(ClickerContext);
}

type Props = {
    children: ReactNode;
};

export const ClickerProvider = ({ children }: Props) => {

    const [state, dispatch] = useReducer(clickerReducer, initialClickerState);

    const providerState = {
        state,
        dispatch
    }

    return (
        <ClickerContext.Provider value={providerState}>
            {children}
        </ClickerContext.Provider>
    )
}
