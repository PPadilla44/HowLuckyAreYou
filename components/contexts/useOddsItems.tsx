import React, { createContext, Dispatch, ReactNode, useContext, useReducer } from 'react';
import { Action,  } from '../../store/oddsItems';
import oddsItemsReducer from "../../store/oddsItems"
import { OddsItemInterface } from '../../types';

const OddsItems = createContext<{ state: OddsItemInterface[], dispatch?: Dispatch<Action>  }>({ state: [] });

export const useOddsItems = () => {
    return useContext(OddsItems);
}

type Props = {
    children: ReactNode;
};

export const OddsItemsProvider = ({ children }: Props) => {

    const [state, dispatch] = useReducer(oddsItemsReducer, [] );

    const providerState = {
        state,
        dispatch
    }

    return (
        <OddsItems.Provider value={providerState}>
            {children}
        </OddsItems.Provider>
    )
}
