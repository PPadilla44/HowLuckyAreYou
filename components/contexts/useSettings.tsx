import { createContext, Dispatch, ReactNode, useContext, useReducer } from "react";
import { Appearance } from "react-native";
import settingsReducer from "../../store/settings"
import { Action, initialSettingsState, SettingsState } from "../../store/settings";

const Settings = createContext<{ state: SettingsState , dispatch?: Dispatch<Action> }>( { state: initialSettingsState } );

export const useSettings = () => {
    return useContext(Settings);
}


type Props = {
    children: ReactNode;
};

export const SettingsProvider = ({ children }: Props) => {

    const [state, dispatch] = useReducer(settingsReducer, initialSettingsState );

    if(state.data.appearance.fromDevice) {
        state.data.appearance.appearance = Appearance.getColorScheme() as "light" | "dark";
    }

    const providerState = {
        state,
        dispatch
    }

    return (
        <Settings.Provider value={providerState}>
            {children}
        </Settings.Provider>
    )
}
