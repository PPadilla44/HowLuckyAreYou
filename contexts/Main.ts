import { useReducer } from "react";
import { View } from "react-native";
import CustomContext from "./context";

export interface MainState {
    count: number;
    odds: number;
    didHit: boolean;
    results: {
        BtnColor: string;
        text: string;
    };
}

export enum MainActionKind {
    INCREASE = "INCREASE",
}

export interface Action {
    type: MainActionKind,
    payload?: any
}

export const initialMainState = {
    count: 0,
    odds: .1,
    didHit: false,
    results: {
        BtnColor: "",
        text: ""
    }
}

function increment(state: MainState) {
    console.log("****************");

    const tempCount = state.count + 1;

    const max = 1 / (state.odds);

    const randomNumFromMax = Math.ceil(Math.random() * max);

    console.log(randomNumFromMax, max);

    if (randomNumFromMax === max) {
        console.log("GOT IT");
        const userOdds = tempCount / max;
        let tempResults
        if (userOdds < 1) {
            tempResults = { BtnColor: "green", text: "YOU ARE REALLY LUCKY" };
        } else if (userOdds > 1) {
            tempResults = { BtnColor: "red", text: "YOU ARE NOT LUCKY" };
        } else {
            tempResults = { BtnColor: "blue", text: "YOU ARE normal" };
        }
        const newState = {
            ...state,
            count: tempCount,
            results: tempResults,
            didHit: true
        }
        console.log(newState);

        return newState
    }

    return {
        ...state,
        count: tempCount
    }
}

export function mainReducer(state: MainState = initialMainState, action: Action) {
    const { type, payload } = action;
    switch (type) {
        case MainActionKind.INCREASE:
            return increment(state)

        default:
            return state;
    }
}

// const ParentComponent = ({ children }) => {

//     const [userState, usersDispatch] = useReducer(mainReducer, initialMainState);

//     const providerState = {
//         userState,
//         usersDispatch
//     }

//     return (
//         <View>
//         </View>
//     )
// }

// export default ParentComponent;