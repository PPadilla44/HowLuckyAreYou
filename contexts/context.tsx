import React, { createContext, useContext } from 'react';

const CustomContext = createContext(null);

export function useCustomContext() {
    return useContext(CustomContext);
}

export default CustomContext;