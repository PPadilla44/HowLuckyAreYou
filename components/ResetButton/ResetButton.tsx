import { Alert, StyleSheet } from 'react-native';
import { Button } from "react-native-elements";
import { View } from "../../components/Themed";
import React from 'react';
import { useClicker } from '../contexts/useClicker';

const ResetButton = () => {

    const { state, dispatch } = useClicker();

    const { didHit } = state;

    const resetAll = () => dispatch!({ type: 'RESET' })

    const createTwoButtonAlert = () => (

        Alert.alert(
            "Reset",
            "Are you sure you would like to reset?",
            [
                {
                    text: "No",
                    onPress: () => console.log("Cancel Pressed"),
                    style: "cancel"
                },
                { text: "Yes", style: "destructive", onPress: resetAll }
            ]
        ));

    return (
        <View testID='resetWrapper' style={[styles.btn, { opacity: didHit ? 1 : 0.5 }]}>
            <Button
                testID='resetBtn'
                title='Reset'
                onPress={createTwoButtonAlert}
                titleStyle={{ color: didHit ? "#830909" : "gray", fontFamily: "Futura" }}
                buttonStyle={{ backgroundColor: undefined }}
            />
        </View>
    )
}

export default ResetButton

const styles = StyleSheet.create({
    btn: {
        position: "absolute",
        left: 10,
        bottom: 10,
    }
})