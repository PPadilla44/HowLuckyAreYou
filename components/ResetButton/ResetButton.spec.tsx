import { render, RenderAPI, fireEvent, act, cleanup } from "@testing-library/react-native";
import React from "react";
import { Alert } from "react-native";
import { ClickerProvider } from "../contexts/useClicker";
import ResetButton from "./ResetButton";

describe('<ResetButton />', () => {

    let wrapper: RenderAPI;

    beforeEach(() => {
        wrapper = render(
            <ClickerProvider>
                <ResetButton />
            </ClickerProvider>
        );
    });

    afterEach(cleanup)


    it('should render correctly', () => {
        wrapper.getByText("Reset");
        expect(wrapper.container.children).toHaveLength(1);
    });

    it('should render didHit FALSE state styles correctly', () => {
        expect(wrapper.getByTestId("resetWrapper")).toHaveStyle({ opacity: 0.5 });
        expect(wrapper.getByText("Reset")).toHaveStyle({ color: "gray" });
    });

    it('should show alert correctly on press', () => {
        const alert = jest.spyOn(Alert, "alert");
        fireEvent.press(wrapper.getByTestId("resetBtn"));
        expect(alert).toHaveBeenCalled();
        expect(alert.mock.calls[0]).toHaveLength(3)
        alert.mockClear()
    });


    it('should call and reset color when confirm reset it pressed', () => {
        const alert = jest.spyOn(Alert, "alert");
        fireEvent.press(wrapper.getByTestId("resetBtn"));

        act(() => {
            alert?.mock?.calls?.[0]?.[2]?.[1].onPress!();
        });
        expect(wrapper.getByTestId("resetWrapper")).toHaveStyle({ opacity: 0.5 });
        expect(wrapper.getByText("Reset")).toHaveStyle({ color: "gray" });
        alert.mockClear()
    });

    it("canceles alert when reset button is pressed and canceled", () => {
        const alert = jest.spyOn(Alert, "alert");
        const logging = jest.spyOn(console, "log")
        const resetBtn = wrapper.getByTestId("resetBtn");

        fireEvent.press(resetBtn);
        expect(alert).toHaveBeenCalled();

        act(() => {
            alert?.mock?.calls?.[0]?.[2]?.[0].onPress!();
        });
        expect(logging).toBeCalled()

    });

    it('should match snapshot', () => {
        expect(wrapper.toJSON()).toMatchSnapshot();
    });


});