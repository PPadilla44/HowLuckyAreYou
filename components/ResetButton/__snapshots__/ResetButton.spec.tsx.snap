// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResetButton /> should match snapshot 1`] = `
<View
  style={
    Array [
      Object {
        "backgroundColor": "black",
        "shadowColor": "white",
      },
      Array [
        Object {
          "bottom": 10,
          "left": 10,
          "position": "absolute",
        },
        Object {
          "opacity": 0.5,
        },
      ],
    ]
  }
  testID="resetWrapper"
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        undefined,
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": false,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID="resetBtn"
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": undefined,
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "gray",
              "fontFamily": "Futura",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Reset
        </Text>
      </View>
    </View>
  </View>
</View>
`;
